utils/big_decimal.cc

```diff
@@ -143,6 +143,97 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
     return x.compare(y) <=> 0;
 }
 
+std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+{
+    if (_scale == other._scale) {
+        return _unscaled_value.compare(other._unscaled_value) <=> 0;
+    }
+
+    // boost::multiprecision::sign() returns -1, 0 or 1
+    const int sign = boost::multiprecision::sign(_unscaled_value);
+    const int sign_other = boost::multiprecision::sign(other._unscaled_value);
+    if (sign != sign_other) {
+        return sign <=> sign_other;
+    }
+    // At this point we know the two signs are equal, so if sign == 0, both signs
+    // and consequently both numbers are zeros.
+    if (sign == 0) {
+        return std::strong_ordering::equal;
+    }
+
+    // At this point we know that both numbers have the same sign, so if one is negative, the other is too.
+    const bool comparing_negatives = (sign < 0);
+
+    // At this point we know that the numbers:
+    // * have different scale
+    // * have the same sign
+    // * neither is zero
+    //
+    // The numbers have the form:
+    //
+    //      auto number = _unscaled_value * std::pow(10, -_scale);
+    //      auto number_other = other._unscaled_value * std::pow(10, -other._scale);
+    //
+    // To compare them, we make the unscaled values the same (or close), so we can directly compare the scales.
+    // To do that we want to compute diff_scale_2:
+    //
+    //      auto unscaled_diff = _unscaled_value / other._unscaled_value;
+    //      auto diff_scale_2 = log2(unscaled_diff);
```

Comment:
In the explanation above, "scale" was about base-10 logarithms, not base-2. Below you switch to log2 for efficiency, but I think this switch is premature in this line of the comment.

Reply:
Renamed to `unscaled_ratio_log2` in v5.